lib/libyui_client/widgets/tree.rb

```diff
@@ -0,0 +1,160 @@
+      #        ],
+      #        "icon_name": "icon",
+      #        "label": "node1",
+      #      },
+      #      {
+      #        "icon_name": "icon",
+      #        "label": "node2"
+      #      }
+      #    ],
+      #    "items_count": 2,
+      #    "label": "node_0",
+      #    "notify": true,
+      #    "vstretch": true
+      #  }
+      # @example
+      #   app.tree(id: 'test_id').items
+      #   # node1
+      #   # node1|node1_1
+      #   # node1|node1_2
+      #   # node1|node1_2|node1_2_1
+      #   # node1|node1_2|node1_2_2
+      #   # node2
+      def items
+        traversed_nodes = []
+        traverse_tree(property(:items), traversed_nodes: traversed_nodes)
+        traversed_nodes
+      end
+
+      # Sends action to select an item from the tree in UI.
+      # @param item [String] value to select from items.
+      # @example Select tree node 'test_node' from tree with id 'test_id'
+      #   app.tree(id: 'test_id').select('test_node')
+      def select(item)
+        action(action: Actions::SELECT, value: item)
+        self
+      end
+
+      # Returns the item currently selected/highlighted in the tree. Item
+      # is represented by its path from root node to leaf in the tree,
+      # in other words, a list of nodes separated by special simbol '|'
+      # @return [String] item selected/highlighted in the tree
+      # @example Get selected item from tree with id "test_id"
+      #  {
+      #    "class": "YTree",
+      #    "debug_label": "node_0",
+      #    "hstretch": true,
+      #    "hweight": 30,
+      #    "icon_base_path": "",
+      #    "id": "test_id",
+      #    "items": [
+      #      {
+      #        "children": [
+      #          {
+      #            "icon_name": "icon",
+      #            "label": "node1_1"
+      #          },
+      #          {
+      #            "children": [
+      #              {
+      #                "label": "node_1_2_1"
+      #              },
+      #              {
+      #                "label": "node_1_2_2",
+      #                "selected": true
+      #              }
+      #            ],
+      #            "icon_name": "icon",
+      #            "label": "node1_2"
+      #          }
+      #        ],
+      #        "icon_name": "icon",
+      #        "label": "node1",
+      #      },
+      #      {
+      #        "icon_name": "icon",
+      #        "label": "node2"
+      #      }
+      #    ],
+      #    "items_count": 2,
+      #    "label": "node_0",
+      #    "notify": true,
+      #    "vstretch": true
+      #  }
+      # @example
+      #   app.tree(id: 'test_id').selected_item
+      #   # node1|node1_2|node1_2_2
+      def selected_item
+        traversed_nodes = []
+        traverse_tree(property(:items), traversed_nodes: traversed_nodes, search_mode: true)
+        traversed_nodes.last
+      end
+
+      private
+
+      def traverse_tree(items, current_path = '', traversed_nodes: [], search_mode: false)
```

Comment:
I would actually propose to create separate structure for TreeItem instead of complex string parsing, as we then store item as a path too.
Can you also you share role of `search_mode` parameter? Looks like we should rely on path instead and return in case last node was processed.

Reply:
Treeitem I guess it would be just a string with node separated by pipes, there is not complex parsing IMO, just concatenation during the process of traversing the tree, because we don't want only the leaves, we want also those intermediate ones, basically all nodes. I didn't go for that direction because I thought those methods should consume themselves, so if we select with a string with pipes, we get items with a string with pipes too.

Perhaps if you have some concrete proposal, I bet this kind be done in many ways :) it was quite tricky for me to do this part and also reuse it for both methods 'items` and `items_selected`. Also we can create a task for this specific thing due it might be time consuming.

`search_mode` traverse the tree and search for an element, the selected one, so basically enable to return earlier instead of going through all the paths, so one we hit the selected one, we exist recursively and we just get the last element accumulated in `item_selected`.
I have it all in one function but rubocop complains about size of function, but it is basically one function with some split for that.

Reply:
I had second thought about my current implementation and I changed, basically I opt for a couple of private methods (replaceable any time) that recursively traverse the tree and they are way easier to maintain that what I had previously, they have less args and still exit early, in ruby for this kind of operations we will need to learn more and more about functional approach, specially in this framework that at the moment does not store offline anything, so it might not be the best choice to create intermediate objects for those parse/traverse process unless we wanted to do something with the structure afterwards.